will/xmpp_plugins.py

```diff
@@ -0,0 +1,289 @@
+import base64
+import ssl
+import logging
+
+from sleekxmpp.util import bytes, sasl
+from sleekxmpp.util.stringprep_profiles import StringPrepError
+from sleekxmpp.stanza import StreamFeatures
+from sleekxmpp.xmlstream import StanzaBase, RestartStream, register_stanza_plugin
+from sleekxmpp.plugins import BasePlugin
+from sleekxmpp.xmlstream.matcher import MatchXPath
+from sleekxmpp.xmlstream.handler import Callback
+from sleekxmpp.features.feature_mechanisms import stanza
+import settings
+
+log = logging.getLogger(__name__)
+
+
+class CustomAuth(stanza.Auth):
+    """
+    """
+
+    name = 'auth'
+    # namespace = 'urn:ietf:params:xml:ns:xmpp-sasl'
+    namespace = 'http://hipchat.com'
+    interfaces = set(('mechanism', 'value'))
+    plugin_attrib = name
+
+    #: Some SASL mechs require sending values as is,
+    #: without converting base64.
+    plain_mechs = set(['X-MESSENGER-OAUTH2'])
+
+    def setup(self, xml):
+        StanzaBase.setup(self, xml)
+        self.xml.tag = self.tag_name()
+        self.xml.set("oauth2_token", "true")
+
+    def get_value(self):
+        if not self['mechanism'] in self.plain_mechs:
+            return base64.b64decode(bytes(self.xml.text))
+        else:
+            return self.xml.text
+
+    def set_value(self, values):
+        if not self['mechanism'] in self.plain_mechs:
+            values = "0%s0%s0will" % (
```

Comment:
@dougkeen premature celebration, sadly.

The snag (as far as I can work out) is that somehow, sleekxmpp is still running the default success handler, which blows up because the hipchat oauth response doesn't have a value (being a self-closing tag.)

Here's my current output

```
DEBUG:sleekxmpp.xmlstream.xmlstream:SEND (IMMED): <auth xmlns="http://hipchat.com" oauth2_token="true" mechanism="PLAIN">TOKENHERE</auth>
DEBUG:sleekxmpp.xmlstream.xmlstream:RECV: <success xmlns="urn:ietf:params:xml:ns:xmpp-sasl" jid="12345_12345@chat.hipchat.com/will||proxy|pubproxy-c5.hipchat.com|5232" api_host="api.hipchat.com" oauth2_token="VALIDTOKENCOMESBACK" chat_host="chat.hipchat.com" muc_host="conf.hipchat.com" web_host="www.hipchat.com" />
self
ERROR:sleekxmpp.xmlstream.xmlstream:Connection error.
ERROR:sleekxmpp.basexmpp:'NoneType' object has no attribute 'process'
Traceback (most recent call last):
  File "~/.virtualenvs/will-core/lib/python2.7/site-packages/sleekxmpp/xmlstream/xmlstream.py", line 1481, in _process
    if not self.__read_xml():
  File "~/.virtualenvs/will-core/lib/python2.7/site-packages/sleekxmpp/xmlstream/xmlstream.py", line 1553, in __read_xml
    self.__spawn_event(xml)
  File "~/.virtualenvs/will-core/lib/python2.7/site-packages/sleekxmpp/xmlstream/xmlstream.py", line 1621, in __spawn_event
    handler.prerun(stanza_copy)
  File "~/.virtualenvs/will-core/lib/python2.7/site-packages/sleekxmpp/xmlstream/handler/callback.py", line 64, in prerun
    self.run(payload, True)
  File "~/.virtualenvs/will-core/lib/python2.7/site-packages/sleekxmpp/xmlstream/handler/callback.py", line 76, in run
    self._pointer(payload)
  File "~/.virtualenvs/will-core/lib/python2.7/site-packages/sleekxmpp/features/feature_mechanisms/mechanisms.py", line 229, in _handle_success
    final = self.mech.process(stanza['value'])
AttributeError: 'NoneType' object has no attribute 'process'
```

More broadly, I'm a bit confused on the auth flow.  From what I can tell, I need three pieces of information to authenticate the user, in this flow
- Sign-in email address  (will-bot-name@gmail.com)
- Password
- Hipchat JabberID.

I believe I need the JID and password to initialize sleek (using email there, even when specifying the server doesn't work), and then the email/pass combo to get the OAuth token.

That seems wonky enough to me that it makes me think I'm doing this wrong - there's gotta be an easier way.

Thoughts?  Thanks again for the help.

Also, if you have any time overlap with Bangkok time (GMT+7) and there's a chat we can hop on, that's great by me.  Happy to flex around your availability.